service/history/task/processor.go

```diff
@@ -135,45 +155,63 @@ func (p *processorImpl) StopShardProcessor(
 func (p *processorImpl) Submit(
 	task Task,
 ) error {
-	scheduler, err := p.prepareSubmit(task)
+	if err := p.priorityAssigner.Assign(task); err != nil {
+		return err
+	}
+
+	submitted, err := p.scheduler.TrySubmit(task)
+	if err != nil {
+		return err
+	}
+
+	if submitted {
+		return nil
+	}
+
+	shardScheduler, err := p.getOrCreateShardTaskScheduler(task.GetShard())
```

Comment:
If we have shard level config, then we can submit to the shard scheduler first then try host-level scheduler. In this case, if there is a hot shard, we can give this shard more resources without impact other shards.

Reply:
Synced offline. Basic idea is that if we have a hot shard, we should throttled it instead of giving it more resources as that will other low throughput domains. By submitting to host level pool, tasks from the hot shard will be assigned a low priority and won't affect other domains.